### ISSUE TITLE
xbox gamepad support 

### ISSUE THREAD

**Issue by iHateInventNames** [minetest/minetest#4153]
It would be nice to have gamepad support like in minecraft.

**Comment by tobyplowy** [comment]
Yes this would be amazing.

**Comment by est31** [comment]
Might require irrlicht changes.

**Comment by rubenwardy** [comment]
Don't tag that label without first checking. It should be possible without modifying irrlicht, even if irrlicht would be the best place.

**Comment by benrob0329** [comment]
https://github.com/jkuhlmann/gainput

**Comment by est31** [comment]
If irrlicht doesn't make it possible there is no "proper" way in getting input: we rely on irrlicht to hold the window handle, and to handle input events. If irrlicht doesn't pass on joystick / gamepad events to the user, we can't do anything about it except ugly hacks like opening a second window that we handle.

But it seems there is [`EET_JOYSTICK_INPUT_EVENT`](http://irrlicht.sourceforge.net/docu/namespaceirr.html#ac9eed96e06e85ce3c86fcbbbe9e48a0cac81558e4607ad260e96ae0f7b889e9a5) so it seems there is support by irrlicht for this. Perhaps I'll go buy a gamepad at monday and I can look into implementing this.

**Comment by sofar** [comment]
I have a logitech f710 which should be compatible (it has the X/D switch to switch modes even) and will test.

**Comment by est31** [comment]
Hrmm, looking at the irrlicht support for gamepads, its not that shiny:

* The D-Pad doesn't work for me, as it is represented as axis, and on axis with indices 6 and 7, while irrlicht only matches axis with indices from 0 up to 5
* It doesn't really support plug & play. If you plug in a gamepad while minetest is running, it won't work. You can do nasty tricks, but you won't get real plug and play working without steadily creating more and more load.

So I've thought of adding irrlicht gamepad support by default, and maybe some optional library that supports gamepads better.
So I'm looking for libraries. One problem with most libraries is that they are big monolithic chunks and manage 200 different things apart from gamepads.

My current favourite is https://github.com/ThemsAllTook/libstem_gamepad . It features a very simple interface, supports plug + play, and I hope that it also supports the D-Pad.

`gainput` which was mentioned by @benrob0329 seems to be much bigger than libstem_gamepad, which is bad as I am not sure whether I can use it without opening a window and stuff. I am not really confident that it supports plug + play.

These two libraries don't seem to support force feedback, which is obviously a minus point... I would like to add force feedback. :(

Of course, one could use SDL, it has good industry support, and is well tested, but its just a giant piece of library, and I don't know either if you can use the gamepad support without having to open a window. Haven't looked a the apis yet. I've read somewhere that SDL supports force feedback.

**Comment by benrob0329** [comment]
What about using the input components of SFML? It's Zlib licensed so those files could potentially even be dropped into MT itself.

**Comment by est31** [comment]
BTW does minecraft really have gamepad support? All I have found was gamepad mods, no native gamepad support on the desktop. Or am I overlooking something?

**Comment by sofar** [comment]
I get no response in vanilla MC 1.9 with a gamepad, so, I assume that it requires modding.

**Comment by est31** [comment]
Cool, then we'll get a feature natively that MC requires mods for.

**Comment by mdoege** [comment]
@est31 MC Windows 10 edition has gamepad support.

**Comment by nolsen42** [comment]
What about playstation controllers too?

**Comment by DonBatman** [comment]
Has Sony released pc drivers?

**Comment by est31** [comment]
@nolsen42 I think it shouldn't be a big difference whether its a sony controller or an xbox one. There is partial support already in minetest master (not 0.4.14), you can try it out by setting the "Client -> Controls -> Enable Joysticks" option and then restarting minetest, with your controller connected. Its not "plug + play" unfortunately (irrlicht's fault not ours); the controller needs to be connected when minetest starts.

**Comment by DanHolli123** [comment]
you can use scptoolkit for sony duelshock 3&4s

**Comment by est31** [comment]
As the person who added joystick support to minetest, they obviously do work great, although there is some room for improvement (see below). I've been latest master (2d1fca51e975a01f84fa3fd9b) and jump works for me as well, so it must be something caused by your modifications @rubenwardy . I've got a nacon `GC-100XF`, and it might have some different mapping.

![gc-100xf_01-250x250](https://cloud.githubusercontent.com/assets/8872119/23325825/a19904d0-faf8-11e6-8b1a-9f4f6979e5f3.png)

For this joystick, the current controls look like (just imagine the nacon image below):

![500px-xbox_controller](https://cloud.githubusercontent.com/assets/8872119/23326123/1b72997c-fafb-11e6-8b44-31dcd7396f8d.png)

So :
```
LB = Left click/Dig/Punch (same as for mouse)
LT = Item selection, leftward
RT = Item selection, rightward
RB = Right click/Place/Interact (same as for mouse)
Left joystick = Movement. Note that they also press "virtual" keys that are being sent to the server for vessel movement.
Right joystick = Looking
A/Green/1 = Jump
B/Red/2 = Fast (called SPECIAL_1)
X/Blue/3 = Sneak
Y/Yellow/4 = Modifier 4
Back = Same as Esc key on desktop
Start = Modifier Start
Modifier 4 + Fast = drop
Modifier 4 + LB = Item selection, leftward (I think I've added this for joysticks that don't have LT)
Modifier 4 + RB = Item selection, rightward (I think I've added this for joysticks that don't have RT)
```

Its been made so that you can easily play with it. E.g. many people want to be able to change their look while sneaking so that they can build bridges. Right now its possible if you use your left hand to control the right joystick, the right hand for the sneak button.

Also you may want to jump while sneaking, and jump when in fast mode. Also note that SPECIAL_1 is sent to servers so it might mean some stuff there too.

And lastly, you may want to place while sneaking when wanting to build near stuff with formspecs.


Note there is no way to "press" joysticks for my controller, they can only be moved around.
D-Pad is currently unused.

@rubenwardy 

> Buttons are pressed twice quite a lot

When you don't like how some buttons are pressed multiple times, you can try setting `repeat_joystick_button_time`. Its very similar to the desktop, if you press esc for some time, the pause menu flashes on and off as well. And this repeat time also directly influences how fast you can select items. Maybe though it can be improved by having two times: one for the first repeat, one for all following repeats. The first repeat time is a bit bigger, so esc won't disappear that fast, but the following repeats are smaller, helping with fast item selection.

> B/Back shouldn't pause the game

About Back exiting formspecs: this was intentional so that you can at least close a chest formspec when you accidentially opened it. Also, formspec support for gamepads was planned, and you need some button to close it. Its the same as on the desktop.

About reconfigurability: sure, it would be great to add! Although I couldn't figure out how to expose different key combos to the players. E.g. you want to be able to sneak and move the same time, but with some keys this should not be possible. Mhh when I'm thinking more of it I guess a very simple processing based on the set key combos could help....

> I can't see where the rotation code is

Search in the src for `JA_FRUSTUM_HORIZONTAL` and `JA_FRUSTUM_VERTICAL`.

So a list of things to improve:

* Reconfigurability by the user
* Formspec support
* Some way to recognize/distinguish different joysticks and providing different (default) mappings for them. May also help with the (quite funny) bugs we had with laptop sensors that made us turn off joysticks by default
* Maybe improve the repeat recognition for buttons
* More buttons, like for screenshots, etc. We need to find out what's useful :)

**Comment by est31** [comment]
See f64914708092958f735337b3a14a33ca4 and #4261, #4221 for the laptop bugs

**Comment by rubenwardy** [comment]
All I changed was the layout , and the debug print showed that the action was being called - so the problem can't be due to my changes (and given that it didn't work at all before...). It must be device specific

**Comment by HummusSamurai** [comment]
This is so cool! Thank you!

**Comment by GreenXenith** [comment]
I hope this is still being worked on, it's a cool PR.

TL;DR the whole thing.

I personally have a PS2 controller with a USB connector that I use for Battlefront II and occasionally Minetest. However, Minetest doesn't do well with my PS2 controller, so instead I use a pad-to-key mapper (QJoyPad) to map my controller to the keyboard and mouse. It does work in formspecs, but if I move a wrong stick it spams my search field :P So builtin support would be nice.
I have an idea for typing using controllers, when in a text field (ie. chat or search) we can change the controller buttons to different functions to either scroll through characters (sticks/dpad for character selection, R/L for shift or something, etc...) or navigate an on-screen keyboard.
Of course, this isn't entirely necessary, just might be nice.
As for formspecs, the same idea could be applied, just change the control scheme for formspec use (new buttons for < > and click and rightclick etc..). Don't know if this has already been suggested or not.
Oh, maybe a diagram in the ESC menu too or something (changes when you re-map)? Might help users.

**Comment by gsantner** [comment]
was (basic) gamepad removed completely?
I'm having here a controller and minetest seems not to care at all about it.
No movement, no click or anything at all on any button.

Minetest 5.1.0 (Linux), Using Irrlicht 1.8.4,  5.1.0-ppa0ubuntu18.04.1 

![grafik](https://user-images.githubusercontent.com/6735650/70193842-5f03b380-16f8-11ea-9828-913fc2ffe8d2.png)

this would be very nice to play on tv screen with controller, but I'm feeling bad to see that it's not working at all (and theres no real developer interest into this as far I can see here).

**Comment by rubenwardy** [comment]
There's a setting to enable it called `enable_joysticks` I believe. The reason it's disabled by default is due to buggy detection which leads to it thinking that some touchpads are joysticks

I'm interested in this, but it's painful to do right and I've been working on other things

**Comment by gsantner** [comment]
Thanks for the hint rubenwardy!
It's not working quite good, but at least I can move now with controller, nice :D 

that's the settings I needed to write at `~/.minetest/minetest.conf`, anywhere on separate lines:
```
enable_joysticks = true
joystick_type = generic
```

**Comment by hecktest** [comment]
I tried using a pad the other day and the camera couldn't stay still, like there was no deadzone to the stick at all.

**Comment by rubenwardy** [comment]
There is a dead zone, however, each gamepads vary massively across model, OS, and drivers. On the same OS, you can find different functionality with the same gamepad!

**Comment by gsantner** [comment]
I had this too. There are 3 profiles - auto,generic,xbox. Xboxhad this issue for me, but generic not.

**Comment by rubenwardy** [comment]
It's worth looking to see if we can find a library which handles detecting the various different gamepad set ups. In the end we may need to read the bindings in from a file

**Comment by gsantner** [comment]
subscribe

**Comment by RQWorldblender** [comment]
Why not try using SDL2 for the window creation and management part? Irrlicht can still be used for rendering all of the graphics.

SuperTuxKart switched to using SDL2 for window creation, but still uses its Irrlicht fork for rendering all of its graphics, in order to gain better game controller support (includes hotplugging). It is possible for Minetest to use SDL2 for the same purpose, but it will require rewriting the window management and related code.

**Comment by rubenwardy** [comment]
> Why not try using SDL2 for the window creation and management part? Irrlicht can still be used for rendering all of the graphics.

This is planned for the next version, we will be shipping with an Irrlicht fork using SDL for window creation and input

**Comment by RQWorldblender** [comment]
> This is planned for the next version, we will be shipping with an Irrlicht fork using SDL for window creation and input

Is this fork another custom one, or is it more like the one the SuperTuxKart is using? What's this fork going to be called if it's not the one from SuperTuxKart?

**Comment by rubenwardy** [comment]
Probably just minetest-irrlicht, users don't really care about the rendering library we use. It doesn't need to be branded

**Comment by RQWorldblender** [comment]
> Probably just minetest-irrlicht, users don't really care about the rendering library we use. It doesn't need to be branded

Okay, that's good enough. In addition to the gamepad support, what about gameplay and being able to navigate the GUI with only a keyboard?

**Comment by netzwanze** [comment]
Hi guys,

I would donate 100€ for the controller / gamepad support.

It would be nice to use mintes also on android tv with controller. Similar to SuperTuxKart.

**Comment by ohhmm** [comment]
> Hi guys,
> 
> I would donate 100€ for the controller / gamepad support.
> 
> It would be nice to use mintes also on android tv with controller. Similar to SuperTuxKart.

Consider using https://www.bountysource.com/ - post bounty, insert this issue browser link.

**Comment by Pxtl** [comment]
Is Android support for joysticks not a thing?   I tried enabling the joystick in "settings" and setting the joystick type to "auto" but no dice.

**Comment by gsantner** [comment]
still not supported (properly)? this is a game that is something I'd expect at last from a game of this type

**Comment by servantoftestator** [comment]
Pxtl I tried using android for joysticks and linux for joysticks and my joysticks did the same thing in both. The input is received and not interpreted by the game right. In my case on android it caused the screen to veer right constantly and I could jump with the wrong button, which is the same thing it did to me on linux.

**Comment by droptile** [comment]
I plugged in a gamepad to my android device and it doesn't work with Minetest but it worked flawlessly with TuxCart. Its unfortunate because Minetest is one of the only FOSS android compatible games that would really benefit from using a gamepad. This issue is over 8 years old ☠️.

**Comment by servantoftestator** [comment]
If you can compile your own droptile there's a PR here https://github.com/minetest/minetest/pull/12888 that would make your gamepad work for gameplay but not formspec menu input on android. You'd still have to use the touchscreen. I actually rebased that PR just so a friend of mine could use their android device to play since mouse right click isn't a input available on android. It was simpler to just use a gamepad with sdl2 input on android then mess around with the touchpad input stuff. Gamepads still need a real fix instead of that hack that doesn't work with formspecs though.

**Comment by Pxtl** [comment]
I haven't looked at the code and it's not my project, but I'd say if the code is clean-enough, "doesn't cover formspec" is good-enough reason to merge.  Partial gamepad support on Android is better than none.  I wouldn't mine reaching up and touching the screen if I'm playing on a handheld android device - obviously that would fail for set-tops but that's probably less common than "phone plus gamepad" or "handheld android".

**Comment by rubenwardy** [comment]
The code is almost ready to merge, we just need to switch to SDL for input (which should be happening soon™)

**Comment by alrito28** [comment]
Hi,

I have downloaded the development version “minetest 5.9.0-dev”. I don't want to put pressure on you, but did you take note of my drawing? :)

https://github.com/minetest/minetest/issues/14496#issue-2209339447

**Comment by rubenwardy** [comment]
We still don't have gamepad support yet, there's no point adding a fancy rebinding interface until that is done

**Comment by Zughy** [comment]
@netzwanze does the bounty still apply?

**Comment by jrom99** [comment]
Hello, I'd like to ask if gamepad support is done by emulating keyboard+mouse keys, or if it supports the gamepad directly.

I ask this because a controller is context dependent, while a keyboard can afford "global" shortcuts (with modifier keys + Esc key) and toggle-like behavior (same key does opposing actions):

In survival mode:

<img width="742" height="422" alt="Image" src="https://github.com/user-attachments/assets/890bc6bf-cdfd-4a0b-a3c6-1083e9c109ea" />

In inventory/crafting mode:

<img width="1698" height="771" alt="Image" src="https://github.com/user-attachments/assets/5a1d67bb-3ee3-4993-871f-ae57c7cbbd54" />

This could be tackled on two fronts:

1. Improved keyboard-only navigation
   - adding more prominent highlight/hover/outline effects would help selecting and moving items in crafting mode
   - actions that are done with the mouse should also allow a single shortcut key or controller button (example: like how both the scroll button or the keys N/B can select items in the hotbar)
2. More action configurability
    - allow for setting shortcuts other than toggle-like behavior
      - Keyboard: I opens the inventory, I (and Esc) can close it
      - XBOX: Y opens inventory, but B closes it 
    - avoid making co-dependent actions
      - Keyboard: Q drops stack, while shift+Q drops single item from stack, but it is configured as "sneak + drop" instead
      - XBOX: this could be dpad down drops stack, while dpad up drops single item

**Comment by DanHolli** [comment]
Gamepad support is not emulating KB/M as doing so would require implementing GP support and a translation layer which can already be done with external tools. This issue is for native support and unless I'm mistaken has already been partially implemented using SDL(1)

**Comment by jrom99** [comment]
I couldn't get my Xbox controller to work with Luanti, so I couldn't properly test those settings, but does it support context dependent keybindings (Y for open inventory/quick move stack) then?

**Comment by y5nw** [comment]
> Hello, I'd like to ask if gamepad support is done by emulating keyboard+mouse keys

No.

That said, it would IMO make sense for gamepads to (eventually) be implemented by adding buttons (and axes, once we have generalized support for analog input) as variants of the `KeyPress` class. This would allow different input modes to share the same code path while (depite the name) remaining separate.

The rest of your post are less relevant to the (direct) goal of adding gamepad support as none of the features suggested are (from a technical perspective) specific to gamepads.

> This issue is for native support and unless I'm mistaken has already been partially implemented using SDL(1)

It is implemented "using SDL" in a relatively loose term in that it currently *does* use SDL, but the information is passed on to (and the current implementation is consequently "made" for) Irrlicht's API which is suboptimal at best (compare the joystick event with it's keyboard or mouse button counterpart). The solution is therefore to rework Irrlicht's API (which, unlike some years ago, we now have control of) to reflect SDL's joystick/gamepad API (not vice versa) or drop Irrlicht as an intermediary here (which requires SDL headers to be available to Luanti, and we should arguably finish the transition to SDL3 at the point), the latter of which is AFAICT the implementation in #12888.

> I couldn't get my Xbox controller to work with Luanti, so I couldn't properly test those settings, but does it support context dependent keybindings (Y for open inventory/quick move stack) then?

No.

**Comment by Soybean44** [comment]
I was curious what is the current state of this? Playing on android with an xbox controller while is an "option" is unplayable right now. Do you guy's plan to get functional controller support working with SDL2 or wait till the transition of SDL3 under the hood before you commit to the transition. If there is a working build where can I find it to test it on my machine?

**Comment by y5nw** [comment]
> If there is a working build where can I find it to test it on my machine?

#12888

**Comment by vvranjek** [comment]
The above PR is closed without merging. Is there any hope we can use gamepads in the future. I would love to play the game, but with my kids the only option is on TV with a gamepad. Playing on a gamepad without any analog input is impossible.

**Comment by NathanSalapat** [comment]
There are tools you can use to remap gamepad input to keyboard so you can use more gamepads and change what the keys do moreso than what you might be able to do solely within Luanti.

**Comment by vvranjek** [comment]
@NathanSalapat this is like saying that you can go buy  fish on the market, drop them in the bathtub, get a fishing pole and there you go, you are fishing! From UX perspective, this is a no go for me, especially because playing something designed to use mouse pointer and moving it around with thumbs is a pain. It's less pain for me to try to implement it myself, which I already did to a degree. I can already navigate the menus with d-pad. Navigating the inventory is my next issue.

**Comment by jrom99** [comment]
If only the inventory also had keyboard-only support (select one item, select stack, discard, general navigation), then it would be easier to remap keys and make it usable with a gamepad.

**Comment by NathanSalapat** [comment]
The software I used let me map a joystick to the mouse so I could do things in formspecs, perfect no, but it got the job done. I never meant to say it was a full solution, but until the solution is added it's the best I know of.

**Comment by jrom99** [comment]
I did something similar in [antimicrox](https://github.com/AntiMicroX/antimicrox-profiles/blob/master/applications/Luanti-Minetest/luanti.gamecontroller.amgp) where I use the d-pad as a slower/more precise mouse in menus and keep the joysticks as regular mouses.

But adding keyboard-only navigation in the menus would still be better for keymapping and controller support. Since https://github.com/luanti-org/luanti/issues/11392, we can at least see which formspec element is selected, but not all of the items are available for keyboard navigation with tab/shift+tab, and it would be useful to navigate using the arrow keys.

<details>
<summary>Current keyboard navigation of the main menu:</summary>
<img width="1082" height="676" alt="Image" src="https://github.com/user-attachments/assets/e400542f-31bf-4169-adef-a6b63bca9f04" />

You can't switch between tabs using the keyboard, so you have to rely on "precise" mouse movement to do so, which is harder to emulate in a gamepad. That said, I'd like if we had RT/RB gamepad support, which could be achieved by just mapping a shortcut (like P/N) to switch between tabs and arrow keys to select elements on menus.
</details>

<details>
<summary>Current keyboard navigation of the paused menu:</summary>
<img width="783" height="407" alt="Image" src="https://github.com/user-attachments/assets/881ac4c1-4ec8-49f8-bae4-5b9f48558c12" />

This one just needs to remove the selectable text label next to the items and maybe add arrow key support and it is ready to go.
</details>

<details>
<summary>Keyboard navigation of the inventory:</summary>
<img width="747" height="684" alt="Image" src="https://github.com/user-attachments/assets/cecc89cc-5999-42e5-9f00-d45b47ff5740" />

This one is trickier since there are mods that affect the inventory, but you can't select all formspec items with tab (you can't select the empty crafting spaces, the hotbar row, regular inventory rows, equipment column).
</details>
